HIP/hip-1195.md

```diff
@@ -0,0 +1,834 @@
+---
+hip: 1195
+title: Hiero hooks and an application to allowances
+author: Michael Tinker <@tinker-michaelj>
+working-group: Richard Bair <@rbair23>, Jasper Potts <@jasperpotts>, Atul Mahamuni <atul@hashgraph.com>,  Matthew DeLorenzo <@littletarzan>, Edward Wertz <@edward-swirldslabs>, Steven Sheehy <@steven-sheehy>
+requested-by: Hashgraph
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+created: 2025-02-19
+discussions-to: https://github.com/hiero-ledger/hiero-improvement-proposals/discussions/1172
+updated: 2025-05-13
+-------------------
+
+## Abstract
+
+We propose **hooks**, programmable Hiero extension points that let users customize the behavior of their entities.
+In principle, hooks could be programmed in any language, but we begin with **EVM hooks**. Users program EVM hooks 
+by writing contracts in a language like Solidity that compiles to EVM bytecode. EVM hooks are either **pure**
+(using neither storage nor external contracts); or **lambdas** (like code running in a cloud's event-driven compute
+offering, which may access a database to use state or call other external services). Users can install multiple
+hooks at different **indexes** on the same entity, until they reach a configured limit.
+
+As a first Hiero extension point, we propose **allowance hooks**. Users can install these hooks on their accounts.
+A Hiero API (HAPI) `CryptoTransfer` transaction can then reference a hook allowance just as it does an ERC-style
+allowance defined in [HIP-376](https://hips.hedera.com/hip/hip-376). The network uses the hook by calling its EVM bytecode at a specific function 
+signature, passing in the details of the transfers proposed in the transaction. If the hook returns `true`, the network 
+continues executing the transfer; otherwise the network rejects the transfer. Installing a hook on an account is 
+analogous to adding a function to a smart contract; the hook executes with the account's privileges when calling 
+Hedera system contracts, just as a smart contract's functions do.
```

Comment:
> the hook executes with the account's privileges when calling Hedera system contracts, just as a smart contract's functions do.

I think we should make the privileges model clearer. My suggestion is to make the hook run with the least privileges possible, like any other smart contract on the network.

Reply:
Added section on privileges model [here](https://github.com/tinker-michaelj/hiero-improvement-proposals/blob/main/HIP/hip-1195.md#privileges-of-the-installing-entity), emphasizing we intend and affirm a model analogous to Ethereum account abstraction, where the executing code acts on behalf of the installer.